### DISCUSSION TITLE
PyScript fetch - thoughts

### DISCUSSION THREAD

**Discussion by FabioRosado** [pyscript/pyscript#1999] [General]
Hello as promised I've been testing pyscript fetch and decided to start this discussion to provide some feedback, I know this is our first interaction and let me start by saying that it makes me very happy to not having to import stuff from outside pyscript and can just add the import to my various `from pyscript` statements.

## Error raising

This one was a bit of a surprise, I never expect that by making a request I could get an exception in python, I would like to be able to handle `!ok` responses by own way without having to either:
- Dot my code with try/excepts
- Write a wrapper to the fetch so it handles exceptions

Now, I understand why we are raising the exception - to chain .json-like methods to the call to fetch so you only use on await. - if we do raise an exception, I would expect to have these two things implemented:
- A custom HTTPException exception
- A way to grab various things from the exception such as status code, text response, possibly headers and anything else (as a side note, this is what boto3 gives you when you get an exception and you can do a lot of handling with this)

## Async flag

This might be a bit of a silly idea, but initially, I forgot to add `async` to the script tag. Could we perhaps just make top-level await work so we don't have to explain to users that they need to do this if they want to use fetch? 

I know that we had various discussions about this, and if I recall correctly, the reason why we decided not to allow top-level async was due to the pyscript code running at unexpected turns instead of a sync-like script. But I was wondering if perhaps we would like to revisit.

## Chaining

I think this one was only unusual to me, given that I am used to use libraries such as aiohttp or httpx (within async), but I was a bit surprised that I could do `await fetch().json()`, not sure if we decided on this to make it easier for new folks to utilize fetch, but given the async nature of making requests in the browser, I'm not sure we should be worried about it because people will wonder what is this magic `await` thing anyway 🤔 

Finally, for context, here's a [very basic pyscript.com example app](https://pyscript-dev.com/@7a3ff64c-85e6-463e-bf92-ce5ed8585f5b/fetch/latest) using fetch and getting errors .

**Comment by WebReflection** [comment]
This is mostly a duplicate of this one: https://github.com/pyscript/pyscript/discussions/1996

Once that lands (I am working on it as we speak) you can have `response = await fetch(...)` and then `await response.json()` which grants you all possible things you want.

The exception already mentions the URL and the status but it doesn't include the text, although, as you said:

> not sure if we decided on this to make it easier for new folks to utilize fetch

absolutely yes! for PoC and RAD nobody wants the double await as in the 90% of the time you just keep writing `fetch(url).then(r => r.json())` in JS and that second callback is harder to explain and deal with due `create_proxy` possible related issues that are even harder to explain and the most frequent issue our users have in general.

`await fetch(url).json()` it is then, if you just care about the resulting object, and it requires zero intermediate callbacks and API explanations.

Once again though, with the other discussion I am splitting optionally anything in two parts so you do what you want and you still get in return python objects instead of JS proxies ... I hope that's welcome.

```python
# currently
result = await fetch(url).json()

# once my changes land
response = await fetch(url)
response.ok
response.status
response.headers
result = await response.json()
```

As example, above code will produce the same result in both cases, the first one is for beginners, the latter one is for people actually knowing more about fetch and how it works and what it returns, but in both cases you will have a Python list or dictionary, not a JS proxy, with JSON, and same goes for bytearray and arrayBuffer.

**Comment by WebReflection** [comment]
btw ... about this ...

> A way to grab various things from the exception such as status code, text response, possibly headers

you can't have response text if you don't `response.text()` even if broken/failed so all this burden should rather be on the advanced use case than in core, imho ... as it's an easy slippery slope for more confusion to me ... if a page is 404 or 500 you probably don't want to bother that server further with a `.text()` completion to know what it says, you just drop the request to me.

It's true that some service returns even `json` as error status but imho that's a stretch of the HTTP status code and I am not sure we should normalize it out of the box.

**Comment by WebReflection** [comment]
closing as this got merged addressing all concerns and allowing `await fetch(...)` a part.
